Sources/SourceKitLSP/Swift/SwiftCodeLensScanner.swift

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import SwiftSyntax
+
+/// Scans a source file for classes or structs annotated with `@main` and returns a code lens for them.
+final class SwiftCodeLensScanner: SyntaxVisitor {
+
+  /// The document snapshot of the syntax tree that is being walked.
+  private var snapshot: DocumentSnapshot
+
+  /// The collection of CodeLenses found in the document.
+  private var result: [CodeLens] = []
+
+  private init(snapshot: DocumentSnapshot) {
+    self.snapshot = snapshot
+    super.init(viewMode: .fixedUp)
+  }
+
+  /// Public entry point. Scans the syntax tree of the given snapshot for an `@main` annotation
+  /// and returns CodeLens's with Commands to run/debug the application.
+  public static func findCodeLenses(
+    in snapshot: DocumentSnapshot,
+    syntaxTreeManager: SyntaxTreeManager
+  ) async -> [CodeLens] {
+    guard snapshot.text.contains("@main") else {
+      // This is intended to filter out files that obviously do not contain an entry point.
+      return []
+    }
+
+    let syntaxTree = await syntaxTreeManager.syntaxTree(for: snapshot)
+    let visitor = SwiftCodeLensScanner(snapshot: snapshot)
+    visitor.walk(syntaxTree)
+    return visitor.result
+  }
+
+  override func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
+    node.attributes.forEach(self.captureLensFromAttribute)
+    return .skipChildren
+  }
+
+  override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+    node.attributes.forEach(self.captureLensFromAttribute)
+    return .skipChildren
+  }
+
+  private func captureLensFromAttribute(attr: AttributeListSyntax.Element) {
+    if attr.trimmedDescription == "@main" {
+      let range = self.snapshot.absolutePositionRange(of: attr.trimmedRange)
+      // TODO: Only provide these code lenses for `.executable` targets.
+      // It doesn't makes sense to provide them for library/plugin targets, even if they have a `@main` attribute.
+      // Until this is addressed `CodeLensTests.testCodeLensRequestWithInvalidProduct` will fail.
+      self.result.append(
+        CodeLens(
+          range: range,
+          command: Command(title: "Run", command: "swift.run", arguments: nil)
```

Comment:
These should be configurable and if configuration is not there then the code lens should be disabled. I think in the #797 (Ignore the title)  we talked about adding the configuration to the `initializationOptions` in VS Code

Reply:
I did read through that thread but was left thinking, are code lenses a special case that need to be turned off in the language server? Other features like Inlay Hints are always provided by sourcekit-lsp and then either enabled or disabled in VS Code with the `editor.inlayHints.enabled` setting.

Similarly VS Code can control this on the UI side with the setting `"editor.codeLens": false`.

Reply:
They might need turned off if the LSP server is being run by a client that doesn't support that code lens

Reply:
If I remember right this is how the java extension works

Reply:
Ah ok I understand, I think this is complementary to my comment below. If the client does command resolution then when it encounters a lens it doesn't know how to handle, it can discard it.